internal/appsec/waf/waf_test.go

```diff
@@ -955,6 +1055,193 @@ func TestEncoder(t *testing.T) {
 	}
 }
 
+// This test needs a working encoder to function properly, as it first encodes the objects before decoding them
+func TestDecoder(t *testing.T) {
+	const intSize = 32 << (^uint(0) >> 63) // copied from recent versions of math.MaxInt
+	const maxInt = 1<<(intSize-1) - 1      // copied from recent versions of math.MaxInt
+	e := encoder{
+		maxDepth:        maxInt,
+		maxStringLength: maxInt,
+		maxArrayLength:  maxInt,
+		maxMapLength:    maxInt,
+	}
+	objBuilder := func(v interface{}) *wafObject {
+		var err error
+		obj := &wafObject{}
+
+		if v, ok := v.(int64); ok {
+			obj.setInt64(toCInt64(int(v)))
+			return obj
+		}
+		if v, ok := v.(uint64); ok {
+			obj.setUint64(toCUint64(uint(v)))
+			return obj
+		}
+		obj, err = e.encode(v)
+		require.NoError(t, err, "Encoding object failed")
+		return obj
+	}
+
+	t.Run("Valid", func(t *testing.T) {
+		for _, tc := range []struct {
+			Name          string
+			Object        *wafObject
+			ExpectedValue interface{}
+		}{
+			{
+				Name:          "string",
+				ExpectedValue: "string",
+				Object:        objBuilder("string"),
+			},
+			{
+				Name:          "Empty string",
+				ExpectedValue: "",
+				Object:        objBuilder(""),
+			},
+			{
+				Name:          "uint64",
+				ExpectedValue: uint64(42),
+				Object:        objBuilder(uint64(42)),
+			},
+			{
+				Name:          "int64",
+				ExpectedValue: int64(42),
+				Object:        objBuilder(int64(42)),
+			},
+			{
+				Name:          "array",
+				ExpectedValue: []interface{}{"str1", "str2", "str3", "str4"},
+				Object:        objBuilder([]string{"str1", "str2", "str3", "str4"}),
+			},
+			{
+				Name:          "Empty array",
+				ExpectedValue: []interface{}{},
+				Object:        objBuilder([]interface{}{}),
+			},
+			{
+				Name:          "struct",
+				ExpectedValue: map[string]interface{}{"Str": "string"},
+				Object: objBuilder(struct {
+					Str string
+				}{Str: "string"}),
+			},
+			{
+				Name:          "Empty struct",
+				ExpectedValue: map[string]interface{}{},
+				Object:        objBuilder(struct{}{}),
+			},
+			{
+				Name:          "map",
+				ExpectedValue: map[string]interface{}{"foo": "bar", "bar": "baz", "baz": "foo"},
+				Object:        objBuilder(map[string]interface{}{"foo": "bar", "bar": "baz", "baz": "foo"}),
+			},
+			{
+				Name:          "Empty map",
+				ExpectedValue: map[string]interface{}{},
+				Object:        objBuilder(map[string]interface{}{}),
+			},
+		} {
+			tc := tc
+			t.Run(tc.Name, func(t *testing.T) {
+				defer free(tc.Object)
+				val, err := decodeObject(tc.Object)
+				require.NoErrorf(t, err, "Error decoding the object: %v", err)
```

Comment:
free if no error:
```suggestion
				val, err := decodeObject(tc.Object)
				require.NoErrorf(t, err, "Error decoding the object: %v", err)
				defer free(tc.Object)
```

Reply:
Why? It seems to me that this risks leaking memory in case of error.
Also I just realised that since all objects are built at the beginning of the test, a test failure will likely result in early termination and some objects not being freed since they haven't been iterated on. I'd argue that it's not a problem since this is a test, but it's good to keep it in mind.

Reply:
Oops my bad I got confused with encode vs decode here.
All good.